hub/codec/codecs.py

```diff
@@ -0,0 +1,380 @@
+        arr = np.frombuffer(decoded_buf, dtype=np.dtype(data["dtype"]))
+        arr = arr.reshape(data["shape"])
+        return arr
+
+
+class Zstd(Base):
+    """Zstd compressor"""
+
+    def __init__(self, level: int):
+        """
+        Initialize Zstd compressor
+
+        Args:
+            level (int): Compression level (1-22).
+        """
+        self.compressor = numcodecs.zstd.Zstd(level)
+        self._msgpack = numcodecs.MsgPack()
+
+    def encode(self, array: np.ndarray) -> bytes:
+        """
+        Encode given array
+
+        Example:
+            arr = np.arange(100, 100, 2, dtype='uint8')
+            arr_encoded = zstd_codec.encode(x)
+
+        Args:
+            array (np.ndarray): Data to be encoded
+
+        Returns:
+            Encoded data.
+        """
+        return self._msgpack.encode(
+            [
+                {
+                    "item": self.compressor.encode(array),
+                    "dtype": array.dtype.name,
+                    "shape": array.shape,
+                }
+            ]
+        )
+
+    def decode(self, bytes_: bytes) -> np.ndarray:
+        """
+        Decode data from buffer.
+
+        Example:
+            arr_decoded = zstd_codec.decode(arr_encoded)
+
+        Args:
+            bytes_ (bytes): Encoded data
+
+        Returns:
+            Decoded data.
+        """
+        data = self._msgpack.decode(bytes_)[0]
+        decoded_buf = self.compressor.decode(data["item"])
+        arr = np.frombuffer(decoded_buf, dtype=np.dtype(data["dtype"]))
+        arr = arr.reshape(data["shape"])
+        return arr
+
+
+class BaseImgCodec(Codec):
+    """Base class for image codecs"""
+
+    def __init__(self, single_channel: bool = True):
+        self.single_channel = single_channel
+        self._msgpack = numcodecs.MsgPack()
+
+    def encode_single_image():
+        return NotImplementedError()
+
+    def decode_single_image():
+        return NotImplementedError()
+
+    def encode(self, arr: np.ndarray):
+        """
+        Encode array with one or multiple images.
+
+        Example:
+            img = np.ones((4, 100, 100, 1))
+            encoding = codec.encode(img)
+
+        Args:
+            array (np.ndarray): Image data to be encoded. Can contain image/s with shapes:
+            (N, M) if image has one channel and single_channel is False,
+            (N, M, 1) if image has one channel and single_channel is True,
+            (N, M, 3).
+            In case of multiple images they should be stacked across the first axis.
+
+        Returns:
+            Encoded dictionary of images and metadata.
+        """
+        append_one = False
+        if self.single_channel and arr.shape[-1] == 1:
+            arr = np.reshape(arr, arr.shape[:-1])
+            append_one = True
+        if not self.single_channel or append_one:
+            shape_dims = 2
+        else:
+            shape_dims = 3
+        assert len(arr.shape) >= shape_dims
```

Comment:
I'm hesitant to allow assertions like this. They usually are better off as test cases, or explicit `raise` statements so we can throw a meaningful error.